src/FSharpVSPowerTools.Core/UntypedAstUtils.fs

```diff
@@ -815,123 +866,183 @@ module Outlining =
             | SynExpr.ForEach (_,_,_,_,_,e,r) ->
-                yield! rcheck Scope.Below  r
+                yield! rcheck Scope.For Collapse.Below  r
                 yield! visitExpr e
             | SynExpr.While (_,_,e,r) ->
-                yield! rcheck Scope.Below  r
+                yield! rcheck Scope.While Collapse.Below  r
                 yield! visitExpr e
             | SynExpr.Lambda (_,_,_,e,r) ->
-                yield! rcheck Scope.Below  r
+                yield! rcheck Scope.Lambda Collapse.``->`` r
                 yield! visitExpr e
             | SynExpr.Lazy (e,r) ->
-                yield! rcheck Scope.Below r
+                yield! rcheck Scope.SpecialFunc Collapse.Below r
                 yield! visitExpr e
             | SynExpr.Quote (_,isRaw,e,_,r) ->
                 // subtract columns so the @@> or @> is not collapsed
-                yield! rcheck Scope.Same <| rangeMod r (if isRaw then 3 else 2) (if isRaw then 3 else 2)
+                yield! rcheck Scope.Quote Collapse.Same <| rangeMod r (if isRaw then 3 else 2) (if isRaw then 3 else 2)
                 yield! visitExpr e
             | SynExpr.Tuple (es,_,r) ->
-                yield! rcheck Scope.Same r
+                yield! rcheck Scope.Tuple Collapse.Same r
                 yield! Seq.collect visitExpr es
-            | SynExpr.Paren(e,_,_,_) ->
+            | SynExpr.Paren (e,_,_,_) ->
                 yield! visitExpr e
-            | SynExpr.Record (_,_,_,r) ->
+            | SynExpr.Record (recCtor,recCopy,recordFields,r) ->
+                if recCtor.IsSome then 
+                    let (_,ctorArgs,_,_,_) = recCtor.Value
+                    yield! visitExpr ctorArgs
+                if recCopy.IsSome then
+                    let (e,_) = recCopy.Value
+                    yield! visitExpr e
+                yield! recordFields |> (Seq.choose(fun(_,e,_)->e)>>Seq.collect visitExpr)
                 // exclude the opening `{` and closing `}` of the record from collapsing
-                yield! rcheck Scope.Same <| rangeMod r 1 1
+                yield! rcheck Scope.Record Collapse.Same <| rangeMod r 1 1
+            | SynExpr.Upcast (e,_,_)  ->
+                yield! visitExpr e
+            | SynExpr.Downcast (e,_,_)  ->
+                yield! visitExpr e
+            | SynExpr.AddressOf(_,e,_,_) ->
+                yield! visitExpr e
+            | SynExpr.InferredDowncast(e,_) ->
+                yield! visitExpr e
+            | SynExpr.InferredUpcast(e,_) ->
+                yield! visitExpr e
+            | SynExpr.DotGet(e,_,_,_) ->
+                yield! visitExpr e
+            | SynExpr.DotSet(e,_,_,_) ->
+                yield! visitExpr e
+            | SynExpr.DotIndexedGet(e,_,_,_) ->
+                yield! visitExpr e
+            | SynExpr.DotIndexedSet(e,_,_,_,_,_) ->
+                yield! visitExpr e
+            | SynExpr.Typed(e,_,_) ->
+                yield! visitExpr e                
```

Comment:
I didn't know that would work! That's pretty sweet. But it's revealed another collapse edge case >.<
![](http://i.imgur.com/zbzerfm.png)